src/libraries/System.Text.Json/tests/System.Text.Json.Tests/Serialization/MetadataTests/DefaultJsonTypeInfoResolverTests.JsonPropertyInfo.cs

```diff
@@ -1178,5 +1178,195 @@ private class MyClassCustomConverterFactory : JsonConverterFactory
+            {
+                TypeInfoResolver = new DefaultJsonTypeInfoResolver()
+                {
+                    Modifiers =
+                    {
+                        jti =>
+                        {
+                            if (jti.Type == typeof(ClassWithExtensionDataAttribute))
+                            {
+                                Assert.Equal(3, jti.Properties.Count);
+                                Assert.True(jti.Properties[2].IsExtensionData);
+                                jti.Properties.RemoveAt(2);
+                            }
+                        }
+                    }
+                }
+            };
+
+            var value = new ClassWithExtensionDataAttribute
+            {
+                Value1 = 1,
+                Value2 = 2,
+                ExtensionData = new Dictionary<string, object>
+                {
+                    ["Value3"] = 3
+                }
+            };
+
+            string json = JsonSerializer.Serialize(value, options);
+            JsonTestHelper.AssertJsonEqual("""{"Value1":1,"Value2":2}""", json);
+
+            value = JsonSerializer.Deserialize<ClassWithExtensionDataAttribute>("""{"unrecognizedValue":42}""", options);
+            Assert.Null(value.ExtensionData);
+        }
+
+        [Theory]
+        [InlineData(typeof(IDictionary<string, object>))]
+        [InlineData(typeof(IDictionary<string, JsonElement>))]
+        [InlineData(typeof(Dictionary<string, object>))]
+        [InlineData(typeof(Dictionary<string, JsonElement>))]
+        [InlineData(typeof(ConcurrentDictionary<string, JsonElement>))]
+        [InlineData(typeof(JsonObject))]
+        public static void EnablingExtensionData_SupportedPropertyType(Type type)
+        {
+            var jti = JsonTypeInfo.CreateJsonTypeInfo(typeof(ClassWithExtensionDataAttribute), new());
+            var jpi = jti.CreateJsonPropertyInfo(type, "ExtensionData");
+
+            Assert.False(jpi.IsExtensionData);
+            jpi.IsExtensionData = true;
+            Assert.True(jpi.IsExtensionData);
+        }
+
+        [Theory]
+        [InlineData(typeof(int))]
+        [InlineData(typeof(string))]
+        [InlineData(typeof(object))]
+        [InlineData(typeof(ClassWithExtensionDataAttribute))]
+        [InlineData(typeof(List<int>))]
+        [InlineData(typeof(IDictionary<string, string>))]
+        [InlineData(typeof(Dictionary<string, int>))]
+        public static void EnablingExtensionData_UnsupportedPropertyType_ThrowsInvalidOperationException(Type type)
+        {
+            var jti = JsonTypeInfo.CreateJsonTypeInfo(typeof(ClassWithExtensionDataAttribute), new());
+            var jpi = jti.CreateJsonPropertyInfo(type, "ExtensionData");
+
+            Assert.False(jpi.IsExtensionData);
+            Assert.Throws<InvalidOperationException>(() => jpi.IsExtensionData = true);
+        }
+
+        public class ClassWithExtensionDataAttribute
+        {
```

Comment:
I'm missing also test where class has no extension data properties and we incorrectly modify contract to make 2 or more

Reply:
Yup, will address in https://github.com/dotnet/runtime/pull/71797#discussion_r916752391

Reply:
It's been already addressed, I'm saying it's the same concern as the one raised in https://github.com/dotnet/runtime/pull/71797#discussion_r916752391. It's been addressed.